runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/StageBundleFactory.java

```diff
@@ -0,0 +1,16 @@
+package org.apache.beam.runners.fnexecution.control;
+
+/**
+ * A bundle factory scoped to a particular
+ * {@link org.apache.beam.runners.core.construction.graph.ExecutableStage}, which has all of the
+ * resources it needs to provide new {@link RemoteBundle RemoteBundles}.
+ *
+ * <p>Closing a StageBundleFactory signals that a a stage has completed and any resources bound to
+ * its lifetime can be cleaned up.
+ */
+public interface StageBundleFactory extends AutoCloseable {
+  /**
+   * Get a new {@link RemoteBundle bundle} for processing the data in an executable stage.
+   */
+  <InputT> RemoteBundle<InputT> getBundle() throws Exception;
```

Comment:
How do clients register bundle output receivers? Note that when interacting with the `SdkHarnessClient` directly, this is done via `newBundle`:

https://github.com/apache/beam/blob/5bb482f536e75bfacac91a546ab3db843dd2a25e/runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClient.java#L114

Note that we cannot register output receivers at operator startup because the output receivers are attached their respective partition contexts (i.e., in the case of Flink outputs are sunk to the `Collector` which can only be used for the lifetime of a partition (`mapPartition` function call)).

Reply:
This is something I had not considered.  This also makes me wonder if stateRequestHandlers can actually be scoped to the operator, or if they need to be scoped to the bundle (which seems to have the lifetime of a partition, not an operator).  At any rate, any partition-scoped resources should probably be passed to `getBundle` by the StageBundleFactory.

Reply:
Agreed. I'm not sure whether the state handler should be per-stage or per-bundle, but output receivers are definitely per-bundle and should probably be passed as something like `Map<String, FnDataReceiver<WindowedValue<?>>>` keyed by PCollection ids of the bundle's outputs.